### PR TITLE
Add toArrayList and toHashMap methods for ReadableArray and ReadableMap. Fixes #4655

### LINKED ISSUE #4655: Ability to dynamically unwrap ReadableMap and ReadableArray
I was working on the android integration of [react-native-intercom](https://github.com/asamiller/react-native-intercom) today and found that I had to forward an arbitrary `ReadableMap` to `Intercom` as a normal `Map` fully unboxed. I quickly realized there was no way to transform the `ReadableMap` into a usable object without recursively deconstructing it by hand.

Ideally, it'd be nice to have the ability to say `unwrap()` or `getObject()`.

My current approach is to use the `ReadableMapKeySetIterator` and [go through each node recursively](https://github.com/ajwhite/react-native-intercom/blob/88f2ddb84d3b850c1999e1f6e941735edd4d3a25/android/src/main/java/com/robinpowered/react/Intercom/IntercomModule.java#L131-L192)

```java

private Map<String, Object> recursivelyDeconstructReadableMap(ReadableMap readableMap) {
    ReadableMapKeySetIterator iterator = readableMap.keySetIterator();
    Map<String, Object> deconstructedMap = new HashMap<>();
    while (iterator.hasNextKey()) {
        String key = iterator.nextKey();
        ReadableType type = readableMap.getType(key);
        switch (type) {
            case Null:
                deconstructedMap.put(key, null);
                break;
            case Boolean:
                deconstructedMap.put(key, readableMap.getBoolean(key));
                break;
            case Number:
                deconstructedMap.put(key, readableMap.getDouble(key));
                break;
            case String:
                deconstructedMap.put(key, readableMap.getString(key));
                break;
            case Map:
                deconstructedMap.put(key, recursivelyDeconstructReadableMap(readableMap.getMap(key)));
                break;
            case Array:
                deconstructedMap.put(key, recursivelyDeconstructReadableArray(readableMap.getArray(key)));
                break;
            default:
                throw new IllegalArgumentException("Could not convert object with key: " + key + ".");
        }

    }
    return deconstructedMap;
}

private List<Object> recursivelyDeconstructReadableArray(ReadableArray readableArray) {
    List<Object> deconstructedList = new ArrayList<>(readableArray.size());
    for (int i = 0; i < readableArray.size(); i++) {
        ReadableType indexType = readableArray.getType(i);
        switch(indexType) {
            case Null:
                deconstructedList.add(i, null);
                break;
            case Boolean:
                deconstructedList.add(i, readableArray.getBoolean(i));
                break;
            case Number:
                deconstructedList.add(i, readableArray.getDouble(i));
                break;
            case String:
                deconstructedList.add(i, readableArray.getString(i));
                break;
            case Map:
                deconstructedList.add(i, recursivelyDeconstructReadableMap(readableArray.getMap(i)));
                break;
            case Array:
                deconstructedList.add(i, recursivelyDeconstructReadableArray(readableArray.getArray(i)));
                break;
            default:
                throw new IllegalArgumentException("Could not convert object at index " + i + ".");
        }
    }
    return deconstructedList;
}
```

This seems like a handy method that could exist on the `ReadableMap` and `ReadableArray` itself.

### PR DESCRIPTION
**Pull request by satya164**
Context #4658

I kept the original commit and author.

cc @mkonicek

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/bridge/ReadableNativeMap.java [hunk 1]
```diff
@@ -48,6 +53,35 @@ public ReadableMapKeySetIterator keySetIterator() {
     return new ReadableNativeMapKeySetIterator(this);
   }
 
+  public HashMap<String, Object>toHashMap() throws ReadableTypeException {
+    ReadableMapKeySetIterator iterator = keySetIterator();
+    HashMap<String, Object> hashMap = new HashMap<>();
+
+    while (iterator.hasNextKey()) {
+      String key = iterator.nextKey();
+      switch (getType(iterator.nextKey())) {
```
Comment (VonD) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/ReadableNativeMap.java]: shouldn't `iterator.nextKey()` be here `key` ? Wouldn't this consume another key ?

File: ReactAndroid/src/main/java/com/facebook/react/bridge/ReadableTypeException.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.facebook.react.bridge;
+
+import javax.annotation.Nullable;
+
+public class ReadableTypeException extends Exception {
+    public ReadableTypeException(String detailMessage) {
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/ReadableTypeException.java]: 2 spaces for indentation in this file

File: ReactAndroid/src/main/java/com/facebook/react/bridge/ReadableNativeMap.java (hunk 1 above)
Comment (satya164) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/ReadableNativeMap.java]: @VonD Oh, I thought I pushed it. Will fix.

**Comment by mkonicek** [comment]
Looks like Krzysztof already approved the original PR.

@facebook-github-bot shipit

**Pull request closed without merging**